sw/device/lib/dif/dif_rv_core_ibex.c

```diff
@@ -0,0 +1,156 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_rv_core_ibex.h"
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/mmio.h"
+
+// #include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "rv_core_ibex_regs.h"
+
+typedef struct ibex_addr_translation_regs {
+  uint32_t ibus_addr_maching;
+  uint32_t ibus_addr_remap;
+  uint32_t ibus_addr_en;
+  uint32_t ibus_reg_en;
+  uint32_t dbus_addr_maching;
+  uint32_t dbus_addr_remap;
+  uint32_t dbus_addr_en;
+  uint32_t dbus_reg_en;
+} ibex_addr_translation_regs_t;
+
+static const ibex_addr_translation_regs_t
+    kRegsMap[kDifRvCoreIbexAddrTranslationSlotCount] = {
+        [kDifRvCoreIbexAddrTranslationSlot_0] =
+            {
+                .ibus_addr_maching =
+                    RV_CORE_IBEX_IBUS_ADDR_MATCHING_0_REG_OFFSET,
+                .ibus_addr_remap = RV_CORE_IBEX_IBUS_REMAP_ADDR_0_REG_OFFSET,
+                .ibus_addr_en = RV_CORE_IBEX_IBUS_ADDR_EN_0_REG_OFFSET,
+                .ibus_reg_en = RV_CORE_IBEX_IBUS_REGWEN_0_REG_OFFSET,
+                .dbus_addr_maching =
+                    RV_CORE_IBEX_DBUS_ADDR_MATCHING_0_REG_OFFSET,
+                .dbus_addr_remap = RV_CORE_IBEX_DBUS_REMAP_ADDR_0_REG_OFFSET,
+                .dbus_addr_en = RV_CORE_IBEX_DBUS_ADDR_EN_0_REG_OFFSET,
+                .dbus_reg_en = RV_CORE_IBEX_DBUS_REGWEN_0_REG_OFFSET,
+            },
+        [kDifRvCoreIbexAddrTranslationSlot_1] =
+            {
+                .ibus_addr_maching =
+                    RV_CORE_IBEX_IBUS_ADDR_MATCHING_1_REG_OFFSET,
+                .ibus_addr_remap = RV_CORE_IBEX_IBUS_REMAP_ADDR_1_REG_OFFSET,
+                .ibus_addr_en = RV_CORE_IBEX_IBUS_ADDR_EN_1_REG_OFFSET,
+                .ibus_reg_en = RV_CORE_IBEX_IBUS_REGWEN_1_REG_OFFSET,
+                .dbus_addr_maching =
+                    RV_CORE_IBEX_DBUS_ADDR_MATCHING_1_REG_OFFSET,
+                .dbus_addr_remap = RV_CORE_IBEX_DBUS_REMAP_ADDR_1_REG_OFFSET,
+                .dbus_addr_en = RV_CORE_IBEX_DBUS_ADDR_EN_1_REG_OFFSET,
+                .dbus_reg_en = RV_CORE_IBEX_DBUS_REGWEN_1_REG_OFFSET,
+            },
+};
+
+/**
+ * Convert the region address and size into a natural power of two address.
+ *
+ * @param addr Region start address.
+ * @param size region size.
+ * @return Napot address
+ */
+static uint32_t to_napot(uint32_t addr, size_t size) {
+  return addr | (size - 1) >> 1;
+}
+
+/**
+ * Split a natural power of two address into a start address and size.
+ *
+ * @param napot Address formated in NAPOT.
+ * @param size  Pointer to receive the region size.
+ * @return The region start address.
+ */
+static uint32_t from_napot(uint32_t napot, size_t *size) {
+  for (size_t i = 1; i < sizeof(uint32_t) * 8u; ++i) {
+    uint32_t ref = (0x01 << i) - 1u;
```

Comment:
I think my advice got missed up here. The thing that needs the `u` prefix is the 1 in the shift: `1u << i`, because `1 << 31` will produce `-1`, which will get sign extended incorrectly.

Reply:
I see, so in that case I would get `-1 - 1 = -2` right?

Reply:
Yes, but only because it's actually `1 << 31 - 1u => -1 - 1u => 0xffff'ffffu - 1u`. Relying on the promotion rules, especially those that require relative ranks, is tricky.